pkg/processor/runtime/rpc/abstract.go

```diff
@@ -148,21 +149,40 @@ func (r *AbstractRuntime) Stop() error {
 		"status", r.GetStatus(),
 		"wrapperProcess", r.wrapperProcess)
 
+	processTerminated := false
+
 	if r.wrapperProcess != nil {
 
-		// stop waiting for process
-		if err := r.processWaiter.Cancel(); err != nil {
-			r.Logger.WarnWith("Failed to cancel process waiting")
+		if r.configuration.ExplicitAckEnabled {
+
+			// send SIGTERM to wrapper process
+			if err := r.signal(syscall.SIGTERM); err != nil {
+				r.Logger.WarnWith("Failed to send termination signal to wrapper process",
+					"pid", r.wrapperProcess.Pid)
+			}
+
+			// wait for process to finish or timeout
+			processTerminated = r.waitForProcessTermination(r.configuration.WorkerTerminationTimeout)
 		}
 
-		r.Logger.WarnWith("Killing wrapper process", "wrapperProcessPid", r.wrapperProcess.Pid)
-		if err := r.wrapperProcess.Kill(); err != nil {
-			r.SetStatus(status.Error)
-			return errors.Wrap(err, "Can't kill wrapper process")
+		if !processTerminated {
+
+			// stop waiting for process
+			if err := r.processWaiter.Cancel(); err != nil {
+				r.Logger.WarnWith("Failed to cancel process waiting")
+			}
```

Comment:
that needs to happen regardless to process has terminated or not

Reply:
So only the `Kill` below it can be skipped in that case

Reply:
`cancel` and `kill` whether the process has been terminated or not
no need to assume `processTerminated`